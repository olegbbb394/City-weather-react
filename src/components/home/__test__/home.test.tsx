import {render, screen, act, logRoles} from "../../../test-utils";
import Home from "../home";
import userEvent from "@testing-library/user-event";
import {getWeatherAction} from "../../../redux/actions/get-weather.action";
import {ICityWeatherItem} from "../../../types/city-weather.model";

describe('Home page', () => {

    it('should render Home page', () => {

        const {container }  = render(<Home/>);
        const heading = screen.getByRole('heading', {name: /city weather/i});
        logRoles(container);
        expect(heading).toBeInTheDocument();
        expect(container).toMatchSnapshot();
    });

    it('should role on Home page', async () => {

        render(<Home/>);

        const getSelect = () => screen.getByRole('button', {name: /City/i});

        expect(getSelect()).toBeInTheDocument();
        expect(screen.queryByText('Kyiv')).not.toBeInTheDocument();
        expect(screen.queryByText('Odesa')).not.toBeInTheDocument();
        expect(screen.queryByText('Kryvyi Rih')).not.toBeInTheDocument();
        expect(screen.queryByText('Dnipro')).not.toBeInTheDocument();

        await act(async () => {
            await userEvent.click(getSelect());
        });

        expect(screen.queryByText('Dnipro')).toBeInTheDocument();
        expect(screen.queryByText('Kyiv')).toBeInTheDocument();
        expect(screen.queryByText('Odesa')).toBeInTheDocument();
        expect(screen.queryByText('Kryvyi Rih')).toBeInTheDocument();
    });

    it('should dispatch name city in getWeatherAction', async () => {
        const dispatchMock = jest.fn();

        const item: ICityWeatherItem | null = {
            id: 703448,
            main: {pressure: 1000, temp: 35},
            name: 'Kyiv',
            weather: [{description: 'good', icon: 'o2d', id: 2, main: 'good'}],
            wind: {speed: 20}
        };

        const select = () => screen.getByLabelText('City');
        const getWeatherActionMock = getWeatherAction(item.name);

        render(<Home/>);

        await act(async () => {
            await userEvent.click(select());
        });

        dispatchMock(getWeatherActionMock);
        expect(select()).toBeInTheDocument();
        expect(dispatchMock).toHaveBeenCalledTimes(1);
    });

    it('should saving  Selected Card data in localStorage', async () => {
        const localStorageSpy = jest.spyOn(window.localStorage.__proto__, 'setItem');

        const item: ICityWeatherItem | null = {
            id: 703448,
            main: {pressure: 1000, temp: 35},
            name: 'Kyiv',
            weather: [{description: 'good', icon: 'o2d', id: 2, main: 'good'}],
            wind: {speed: 20}
        };

        const dispatchMock = jest.fn();
        const thunk = getWeatherAction(item.name);

        render(<Home/>);

        expect(localStorageSpy).toBeTruthy();
        expect(item.name).toBeTruthy();
        dispatchMock(thunk);
        expect(dispatchMock).toHaveBeenCalledTimes(1);
    });
});
