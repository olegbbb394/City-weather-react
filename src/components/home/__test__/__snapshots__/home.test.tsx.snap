// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home page should render Home page 1`] = `
<div>
  <div
    class="MuiContainer-root MuiContainer-maxWidthLg css-h9p1cd-MuiContainer-root"
  >
    <header>
      <h1
        class="MuiTypography-root MuiTypography-h3 css-88618i-MuiTypography-root"
      >
        City weather
      </h1>
    </header>
    <section>
      <div
        class="MuiBox-root css-gqssyb"
      >
        <div
          class="MuiFormControl-root css-1jveyxb-MuiFormControl-root"
        >
          <label
            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-colorSuccess MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined css-150d6dw-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink="false"
            id="selectId"
          >
            City
          </label>
          <div
            class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorSuccess MuiInputBase-formControl  css-1k8trn2-MuiInputBase-root-MuiOutlinedInput-root-MuiSelect-root"
          >
            <div
              aria-expanded="false"
              aria-haspopup="listbox"
              aria-labelledby="labelId selectId"
              class="MuiSelect-select MuiSelect-outlined MuiInputBase-input MuiOutlinedInput-input css-11u53oe-MuiSelect-select-MuiInputBase-input-MuiOutlinedInput-input"
              id="selectId"
              role="button"
              tabindex="0"
            >
              <span
                class="notranslate"
              >
                â€‹
              </span>
            </div>
            <input
              aria-hidden="true"
              aria-invalid="false"
              class="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
              tabindex="-1"
              value=""
            />
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconOutlined css-hfutr2-MuiSvgIcon-root-MuiSelect-icon"
              data-testid="ArrowDropDownIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M7 10l5 5 5-5z"
              />
            </svg>
            <fieldset
              aria-hidden="true"
              class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="css-yjsfm1"
              >
                <span>
                  City
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
      </div>
      <form
        autocomplete="off"
        class="MuiBox-root css-xn6ruc"
        novalidate=""
      />
    </section>
    <section>
      <div
        class="MuiGrid2-root MuiGrid2-container MuiGrid2-direction-xs-row MuiGrid2-spacing-xs-2 css-1p81imu-MuiGrid2-root"
      />
    </section>
  </div>
</div>
`;
